File: sim/land.ts
import { LandState } from './types'

const MAX_NUTRIENTS = 500
const MAX_MOISTURE = 100
const MAX_TEMPERATURE = 100

export class Land {
  id: number
  col: number; row: number
  x: number; y: number
  width: number; height: number

  sunlight: number       // 0–100
  nutrients: number      // 0–500
  moisture: number       // 0–100
  temperature: number    // 0–100
  fertility: number      // 0–1

  nutRegen: number       // nutrients regenerated per second
  midTemp: number        // temperature midpoint for sine oscillation
  tempRadius: number     // amplitude of temperature sine wave

  constructor(id: number, col: number, row: number, x: number, y: number, width: number, height: number) {
    this.id = id
    this.col = col; this.row = row
    this.x = x; this.y = y
    this.width = width; this.height = height

    this.sunlight = 0
    this.nutrients = 0
    this.moisture = 0
    this.temperature = 0
    this.fertility = 0

    this.nutRegen = 5.5
    this.midTemp = 50
    this.tempRadius = 10
  }

  setSunlight(value: number): void {
    this.sunlight = value
  }

  changeNutrients(delta: number): void {
    this.nutrients = Math.max(0, Math.min(MAX_NUTRIENTS, this.nutrients + delta))
  }

  changeMoisture(delta: number): void {
    this.moisture = Math.max(0, Math.min(MAX_MOISTURE, this.moisture + delta))
  }

  updateTemperature(time: number): void {
    const c = 1
    this.temperature = Math.max(0, Math.min(MAX_TEMPERATURE, this.tempRadius * Math.sin(time * c) + this.midTemp))
  }

  updateFertility(): void {
    this.fertility = Math.max(0, Math.min(1, (this.nutrients / MAX_NUTRIENTS + this.moisture / MAX_MOISTURE) / 2))
  }

  pointInLand(px: number, py: number): boolean {
    return px >= this.x && px <= this.x + this.width && py >= this.y && py <= this.y + this.height
  }

  toJSON(): LandState {
    return {
      id: this.id,
      col: this.col, row: this.row,
      x: this.x, y: this.y,
      width: this.width, height: this.height,
      sunlight: this.sunlight,
      nutrients: this.nutrients,
      moisture: this.moisture,
      temperature: this.temperature,
      fertility: this.fertility,
    }
  }
}

File: game_cpp/src/land.cpp
#include "land.h"

land::land(POINT _top_left, int width, int height)
{
	nutrient_level = 0;
	temperature = 0;
	moisture_level = 0;
	fertility = 0;
	sunlight = 0;

	top_left = _top_left;
	land_width = width;
	land_height = height;

	SetRect(&brc, top_left.x, top_left.y, top_left.x + land_width, top_left.y + land_height);

}

land::~land()
{

}

void land::set_sunlight(float value)
{
	sunlight = value;
}

void land::change_nutrients(float delta)
{
	//delta > 0 -> nutrients increase
	//delta < 0 -> nutrients decrease
	nutrient_level += delta;
	if (nutrient_level < 0)
	{
		nutrient_level = 0;
	}
	if (nutrient_level > MAX_NUTRIENTS)
	{
		nutrient_level = MAX_NUTRIENTS;
	}
}

void land::change_moisture(float delta)
{
	//moisture is in percentage.
	//so min: 0 and max: 100.

	moisture_level += delta;
	if (moisture_level < 0)
	{
		moisture_level = 0;
	}
	if (moisture_level > MAX_MOISTURE)
	{
		moisture_level = MAX_MOISTURE;
	}
}

void land::calculate_temperature(float time)
{
	//time must be the time elapsed since the beginning of the simulation, NOT a time difference.

	//Play with c to adjust the speed of temperature oscillation.
	float c = 1;
	temperature = temp_radius*sin(time*c) + mid_temp;
	if (temperature < 0)
	{
		temperature = 0;
	}
	if (temperature > MAX_TEMPERATURE)
	{
		temperature = MAX_TEMPERATURE;
	}
}

void land::calculate_fertility()
{
	//fertility is a summary variable that amounts to an average of moisture, temperature, and nutrient level.
	//divide by 2 because, for now, only taking two factors into account.
	fertility = (nutrient_level / MAX_NUTRIENTS + moisture_level / MAX_MOISTURE) / 2;
	//fertility += temperature / 100;
	
	if (fertility < 0)
	{
		fertility = 0;
	}
	if (fertility > 1)
	{
		fertility = 1;
	}

}

bool land::point_in_land(POINT pt)
{
	return PtInRect(&brc, pt);
}